lib/SPIRV/SPIRVToOCL20.cpp

```diff
@@ -211,350 +97,32 @@ void SPIRVToOCL20::visitCallSPIRVMemoryBarrier(CallInst *CI) {
-}
-
-void SPIRVToOCL20::visitCallSPIRVBuiltin(CallInst *CI, Op OC) {
-  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
-  mutateCallInstOCL(M, CI,
-                    [=](CallInst *, std::vector<Value *> &Args) {
-                      return OCLSPIRVBuiltinMap::rmap(OC);
-                    },
-                    &Attrs);
-}
-
-void SPIRVToOCL20::visitCallSPIRVGroupBuiltin(CallInst *CI, Op OC) {
-  auto DemangledName = OCLSPIRVBuiltinMap::rmap(OC);
-  assert(DemangledName.find(kSPIRVName::GroupPrefix) == 0);
-
-  std::string Prefix = getGroupBuiltinPrefix(CI);
-
-  bool HasGroupOperation = hasGroupOperation(OC);
-  if (!HasGroupOperation) {
-    DemangledName = Prefix + DemangledName;
-  } else {
-    auto GO = getArgAs<spv::GroupOperation>(CI, 1);
-    StringRef Op = DemangledName;
-    Op = Op.drop_front(strlen(kSPIRVName::GroupPrefix));
-    bool Unsigned = Op.front() == 'u';
-    if (!Unsigned)
-      Op = Op.drop_front(1);
-    DemangledName = Prefix + kSPIRVName::GroupPrefix +
-                    SPIRSPIRVGroupOperationMap::rmap(GO) + '_' + Op.str();
-  }
-  assert(CI->getCalledFunction() && "Unexpected indirect call");
-  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
-  mutateCallInstOCL(M, CI,
-                    [=](CallInst *, std::vector<Value *> &Args) {
-                      Args.erase(Args.begin(),
-                                 Args.begin() + (HasGroupOperation ? 2 : 1));
-                      if (OC == OpGroupBroadcast)
-                        expandVector(CI, Args, 1);
-                      return DemangledName;
-                    },
-                    &Attrs);
-}
-
-void SPIRVToOCL20::visitCallSPIRVPipeBuiltin(CallInst *CI, Op OC) {
-  auto DemangledName = OCLSPIRVBuiltinMap::rmap(OC);
-  bool HasScope = DemangledName.find(kSPIRVName::GroupPrefix) == 0;
-  if (HasScope)
-    DemangledName = getGroupBuiltinPrefix(CI) + DemangledName;
-
-  assert(CI->getCalledFunction() && "Unexpected indirect call");
-  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
-  mutateCallInstOCL(
+  CallInst *CiFixedAs = mutateCallInstOCL(
       M, CI,
       [=](CallInst *, std::vector<Value *> &Args) {
-        if (HasScope)
-          Args.erase(Args.begin(), Args.begin() + 1);
-
-        if (!(OC == OpReadPipe || OC == OpWritePipe ||
-              OC == OpReservedReadPipe || OC == OpReservedWritePipe))
-          return DemangledName;
-
-        auto &P = Args[Args.size() - 3];
-        auto T = P->getType();
-        assert(isa<PointerType>(T));
-        auto ET = T->getPointerElementType();
-        if (!ET->isIntegerTy(8) ||
-            T->getPointerAddressSpace() != SPIRAS_Generic) {
-          auto NewTy = PointerType::getInt8PtrTy(*Ctx, SPIRAS_Generic);
-          P = CastInst::CreatePointerBitCastOrAddrSpaceCast(P, NewTy, "", CI);
+        for (size_t I = 0; I < Args.size(); ++I) {
+          Value *PtrArg = Args[I];
+          if (PtrArg->getType()->isPointerTy()) {
```

Comment:
You're using `PtrArg->getType()` a few times, so the code might be more readable if you factor it out into a new variable.

Reply:
Refactored.